internal/pkg/discovery/adss_handler.go

```diff
@@ -0,0 +1,198 @@
+	subIDFmtStr   = `%0` + strconv.Itoa(maxUintDigits) + `d`
+)
+
+// recvFromStream receives discovery requests from the subscriber.
+func (adss *adsServer) recvFromStream(id int64, downstream DiscoveryStream) {
+	log.Infof("Received from stream %d", id)
+	for {
+		discoveryRequest, err := downstream.Recv()
+		if err != nil {
+			log.Errorf("error while recv discovery request from subscriber %s: %v", fmt.Sprintf(subIDFmtStr, id), err)
+			break
+		}
+		log.Infof("Got discovery request from subscriber %s: %v", fmt.Sprintf(subIDFmtStr, id), discoveryRequest)
+		if discoveryRequest.GetVersionInfo() == "" {
+			resources, err := adss.generateResources(discoveryRequest.GetTypeUrl())
+			if err != nil {
+				// TODO: Do not push empty resources if there was an error during resource generation,
+				// because that may cause unintentional removal of the subscribed resources.
+				log.Errorf("failed to generate resources of type %s: %v", discoveryRequest.GetTypeUrl(), err)
+			}
+			log.Infof("Sending initial config snapshot for type %s: %s", discoveryRequest.GetTypeUrl(), resources)
+			if err := sendToStream(downstream, discoveryRequest.GetTypeUrl(), resources, strconv.FormatInt(time.Now().Unix(), 10)); err != nil {
+				log.Errorf("failed to send initial config snapshot for type %s: %v", discoveryRequest.GetTypeUrl(), err)
+			}
+		}
+	}
+}
+
+func (adss *adsServer) generateResources(typeUrl string) ([]*anypb.Any, error) {
+	handler, found := adss.handlers[typeUrl]
+	if !found {
+		return []*anypb.Any{}, nil
+	}
+
+	log.Infof("Generating config snapshot for type %s", typeUrl)
+	resources, err := handler.GenerateResponse()
+	if err != nil {
+		log.Errorf("failed generating resources for type %s: %v", typeUrl, err)
+		return []*anypb.Any{}, fmt.Errorf("failed generating resources for type %s: %w", typeUrl, err)
+	}
+	return resources, nil
+}
+
+// sendToStream sends XDS resources to the subscriber.
+func sendToStream(downstream DiscoveryStream, typeUrl string, xdsResources []*anypb.Any, version string) error {
+	if err := downstream.Send(&discovery.DiscoveryResponse{
+		TypeUrl:     typeUrl,
+		VersionInfo: version,
+		Resources:   xdsResources,
+		ControlPlane: &envoycfgcorev3.ControlPlane{
+			Identifier: os.Getenv("POD_NAME"),
+		},
+		Nonce: version,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (adss *adsServer) subscribersLen() int {
+	length := 0
+	adss.subscribers.Range(func(_, _ interface{}) bool {
+		length++
+		return true
+	})
+	return length
+}
+
+func (adss *adsServer) Push(pushRequest xds.PushRequest) error {
+	if adss.subscribersLen() == 0 {
+		log.Infof("Skip pushing XDS resources for request [type=%s,resources=%v] as there are no subscribers", pushRequest.TypeUrl, pushRequest.Resources)
+		return nil
+	}
+
+	resources := pushRequest.Resources
+	if resources == nil {
+		var err error
+		resources, err = adss.generateResources(pushRequest.TypeUrl)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Readability suggestion avoiding variable reassignment:

```suggestion
	var resources []*anypb.Any
	if pushRequest.Resources == nil {
		var err error
		resources, err = adss.generateResources(pushRequest.TypeUrl)
		if err != nil {
			return err
		}
	} else {
		resources = pushRequest.Resources
	}
```